lib/Sema/ConstraintSystem.h

```diff
@@ -2331,6 +2331,44 @@ class ConstraintSystem {
                           ConstraintLocatorBuilder locator);
 
 public:
+  // Build a disjunction for the choices between an Optional type and
+  // it's underlying type. We'll make the choice of the Optional
+  // preferred, and select that if the expression type-checks
+  // appropriately with that type.
+  void buildDisjunctionForImplicitlyUnwrappedOptional(
+      TypeVariableType *tv, Type type, ConstraintLocator *locator) {
+
+    // Create the constraint to bind to the optional type and make it
+    // the favored choice.
+    auto *bindToOptional =
+        Constraint::create(*this, ConstraintKind::Bind, tv, type, locator);
+    bindToOptional->setFavored();
+
+    Type underlyingType;
+
+    // FIXME: Support unwrapping results of function calls.
+    assert(!type->is<AnyFunctionType>());
+
+    underlyingType = type->getWithoutSpecifierType()->getAnyOptionalObjectType();
+    assert(underlyingType);
+
+    if (type->is<LValueType>())
```

Comment:
Do both come up here? Maybe add comments explaining why/when

Reply:
I suspect `InOut` shouldn't happen so I'll switch that to an assert in a separate commit.